mlir/lib/Dialect/Math/Transforms/CanonicalizeF32Promotion.cpp

```diff
@@ -0,0 +1,73 @@
+//===- CanonicalizeF32Promotion.cpp - Remove redundant extf/truncf pairs
+//----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements removing redundant extf/truncf pairs inserted from
+// LegalizeToF32.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/Math/Transforms/Passes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::math {
+#define GEN_PASS_DEF_MATHCANONICALIZEF32PROMOTION
+#include "mlir/Dialect/Math/Transforms/Passes.h.inc"
+} // namespace mlir::math
+
+using namespace mlir;
+
+namespace {
+
+struct CanonicalizeF32PromotionRewritePattern final
+    : OpRewritePattern<arith::ExtFOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(arith::ExtFOp op,
+                                PatternRewriter &rewriter) const final {
+    if (auto innertruncop = op.getOperand().getDefiningOp<arith::TruncFOp>()) {
```

Comment:
How about the case as below:
```mlir
        %0 = math.absf %arg0 : vector<32xbf16>
        %1 = math.sin %0 : vector<32xbf16>
        %2 = math.cos %0: vector<32xbf16>
        %3 = arith.addf %1, %2
```

Reply:
Included this case as branch case in [canonicalize-f32-promotion.mlir](https://github.com/crazydemo/llvm-project/pull/1/files#diff-23020567471b87d9fa12d9e9be9851417f9abb82d394dfdb4d30328e6655633e). Based on the current upstream status, we will get code like below. bf16 legalization is not applied to `arith.addf`, as the current upstream does not add this op into the `illegal op list`. 
```
func.func @bf16_branch_vector(%arg0: vector<32x32x32xbf16>) -> vector<32x32x32xbf16> {
    %0 = arith.extf %arg0 : vector<32x32x32xbf16> to vector<32x32x32xf32>
    %1 = math.absf %0 : vector<32x32x32xf32>
    %2 = math.sin %1 : vector<32x32x32xf32>
    %3 = arith.truncf %2 : vector<32x32x32xf32> to vector<32x32x32xbf16>
    %4 = math.cos %1 : vector<32x32x32xf32>
    %5 = arith.truncf %4 : vector<32x32x32xf32> to vector<32x32x32xbf16>
    %6 = arith.addf %3, %5 : vector<32x32x32xbf16>
    return %6 : vector<32x32x32xbf16>
}
```
I will have another PR submitted to graph compiler repo, which enables graph compiler specific bf16 legalization pass (has our own illegal op list, and contain the machine info). Graph compiler specific bf16 legalization full process will generate the below ir.
```
func.func @bf16_branch_vector(%arg0: vector<32x32x32xbf16>) -> vector<32x32x32xbf16> {
    %0 = arith.extf %arg0 : vector<32x32x32xbf16> to vector<32x32x32xf32>
    %1 = math.absf %0 : vector<32x32x32xf32>
    %2 = math.sin %1 : vector<32x32x32xf32>
    %3 = math.cos %1 : vector<32x32x32xf32>
    %4 = arith.addf %2, %3 : vector<32x32x32xf32>
    %5 = arith.truncf %4 : vector<32x32x32xf32> to vector<32x32x32xbf16>
    return %5 : vector<32x32x32xbf16>
}
```